src/Hl7.Fhir.Core/Model/ElementDefinitionExtensions.cs

```diff
@@ -144,5 +183,250 @@ public static ElementDefinition WithBinding(this ElementDefinition ed, string va
+        }
+
+        /// <summary>
+        /// Determines whether the cardinality allows repetition of the element.
+        /// </summary>
+        public static bool IsRepeating(this ElementDefinition defn) => defn != null && defn.Max != "1" && defn.Max != "0";
+
+        /// <summary>
+        /// Determines whether the given path navigates to an extension or modifierExtension.
+        /// </summary>
+        public static bool IsExtensionPath(string path) => !string.IsNullOrEmpty(path) && (path.EndsWith(".extension") || path.EndsWith(".modifierExtension"));
+
+        /// <summary>
+        /// Determines whether the given element is the root element of an extension or modifierExtension.
+        /// </summary>
+        public static bool IsExtension(this ElementDefinition defn) => defn != null && IsExtensionPath(defn.Path);
+
+        /// <summary>Determines if the specified element path represents a root element.</summary>
+        public static bool IsRootPath(string path) => !string.IsNullOrEmpty(path) && !path.Contains('.');
+
+        /// <summary>
+        /// Determines if the specified element is the root element.
+        /// </summary>
+        public static bool IsRootElement(this ElementDefinition defn) => defn != null && IsRootPath(defn.Path);
+
+        /// <summary>Returns a list of distinct type codes supported by the specified element definition.</summary>
+        /// <param name="types">A list of element types.</param>
+        /// <returns>A list of type code strings.</returns>
+        public static IReadOnlyList<string> DistinctTypeCodes(this List<ElementDefinition.TypeRefComponent> types)
+            => types.Where(t => t.Code != null).Select(t => t.Code).Distinct().ToList();
+
+        /// <summary>Returns a list of distinct type codes supported by the specified element definition.</summary>
+        /// <param name="elem">An <see cref="ElementDefinition"/> instance.</param>
+        /// <returns>A list of type code strings.</returns>
+        public static IReadOnlyList<string> DistinctTypeCodes(this ElementDefinition elem) => elem.Type.DistinctTypeCodes();
+
+        /// <summary>
+        /// Returns the only element of a sequence,
+        /// or the default value if the sequence is empty or contains multiple elements.
+        /// <para>
+        /// If the specified sequence contains multiple elements, then this method returns the default value.
+        /// The standard LINQ method <see cref="Enumerable.SingleOrDefault{TSource}(IEnumerable{TSource})"/>
+        /// would throw an exception.
+        /// </para>
+        /// </summary>
+        /// <typeparam name="T">The type of the elements of <paramref name="sequence"/></typeparam>
+        /// <param name="sequence">An <see cref="IEnumerable{T}"/> to return the single element of.</param>
+        /// <returns>A value of type <typeparamref name="T"/>.</returns>
+        public static T? SafeSingleOrDefault<T>(this IEnumerable<T> sequence)
```

Comment:
Shouldn't this pretty specific method be `internal`?

Reply:
I've made it private even